plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergUtil.java

```diff
@@ -13,36 +13,170 @@
+        parameters.put(METADATA_LOCATION_PROP, newMetadataLocation);
+        parameters.remove(TRINO_TABLE_METADATA_INFO_VALID_FOR); // no longer valid
+
+        TableInput tableInput = new TableInput()
                 .withName(tableName)
                 .withOwner(owner.orElse(null))
-                .withParameters(ImmutableMap.<String, String>builder()
-                        .putAll(parameters)
-                        .put(TABLE_TYPE_PROP, ICEBERG_TABLE_TYPE_VALUE.toUpperCase(ENGLISH))
-                        .buildKeepingLast())
                 // Iceberg does not distinguish managed and external tables, all tables are treated the same and marked as EXTERNAL
                 .withTableType(EXTERNAL_TABLE.name());
+
+        if (cacheTableMetadata) {
+            // Store table metadata sufficient to answer information_schema.columns and system.metadata.table_comments queries, which are often queried in bulk by e.g. BI tools
+            String comment = metadata.properties().get(TABLE_COMMENT);
+            Optional<List<Column>> glueColumns = glueColumns(typeManager, metadata);
+
+            boolean canPersistComment = (comment == null || comment.length() <= GLUE_TABLE_PARAMETER_LENGTH_LIMIT);
+            boolean canPersistColumnInfo = glueColumns.isPresent();
+            boolean canPersistMetadata = canPersistComment && canPersistColumnInfo;
+
+            if (canPersistMetadata) {
+                tableInput.withStorageDescriptor(new StorageDescriptor()
+                        .withColumns(glueColumns.get()));
+
+                if (comment != null) {
+                    parameters.put(TABLE_COMMENT, comment);
+                }
+                else {
+                    parameters.remove(TABLE_COMMENT);
+                }
+                parameters.put(TRINO_TABLE_METADATA_INFO_VALID_FOR, newMetadataLocation);
+            }
+        }
+
+        tableInput.withParameters(parameters);
+
+        return tableInput;
+    }
+
+    private static Optional<List<Column>> glueColumns(TypeManager typeManager, TableMetadata metadata)
+    {
+        List<Types.NestedField> icebergColumns = metadata.schema().columns();
+        ImmutableList.Builder<Column> glueColumns = builderWithExpectedSize(icebergColumns.size());
+
+        boolean firstColumn = true;
+        for (Types.NestedField icebergColumn : icebergColumns) {
+            String glueTypeString = toGlueTypeStringLossy(icebergColumn.type());
+            if (glueTypeString.length() > GLUE_COLUMN_TYPE_LENGTH_LIMIT) {
+                return Optional.empty();
+            }
+            String trinoTypeName = TypeConverter.toTrinoType(icebergColumn.type(), typeManager).getTypeId().getId();
+            Column column = new Column()
+                    .withName(icebergColumn.name())
+                    .withType(glueTypeString)
+                    .withComment(icebergColumn.doc());
```

Comment:
Previously, this connector could create tables with column comments > 255 characters, but it fails after this change.

Reply:
good catch!